src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -595,14 +696,109 @@ CreateShardsOnWorkers(Oid distributedRelationId, List *shardPlacements,
 
 
 /*
- * WorkerCreateShard applies DDL commands for the given shardId to create the
- * shard on the worker node. Commands are sent to the worker node over the
- * given connection.
+ * RelationShardListForShardCreate gets a shard interval and returns the placement
+ * accesses that would happen when a placement of the shard interval is created.
  */
-void
-WorkerCreateShard(Oid relationId, int shardIndex, uint64 shardId, List *ddlCommandList,
-				  List *foreignConstraintCommandList, MultiConnection *connection)
+static List *
+RelationShardListForShardCreate(ShardInterval *shardInterval)
 {
+	List *relationShardList = NIL;
+	RelationShard *relationShard = NULL;
+	Oid relationId = shardInterval->relationId;
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(relationId);
+	List *referencedRelationList = cacheEntry->referencedRelationsViaForeignKey;
+	List *referencingRelationList = cacheEntry->referencingRelationsViaForeignKey;
+	List *allForeignKeyRelations =
+		list_concat_unique_oid(referencedRelationList, referencingRelationList);
+	int shardIndex = -1;
+
+	/* record the placement access of the shard itself */
+	relationShard = CitusMakeNode(RelationShard);
+	relationShard->relationId = relationId;
+	relationShard->shardId = shardInterval->shardId;
+	relationShardList = list_make1(relationShard);
+
+	if (cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+		cacheEntry->colocationId != INVALID_COLOCATION_ID)
+	{
+		shardIndex = ShardIndex(shardInterval);
+	}
+
+	/* all foregin constraint relations */
+	if (allForeignKeyRelations != NIL)
+	{
+		ListCell *relationIdCell = NULL;
+
+		foreach(relationIdCell, allForeignKeyRelations)
+		{
+			Oid fkeyRelationid = lfirst_oid(relationIdCell);
+			RelationShard *fkeyRelationShard = NULL;
+			uint64 fkeyShardId = INVALID_SHARD_ID;
+
+			if (!IsDistributedTable(fkeyRelationid))
+			{
+				/* we're not interested in local tables */
+				continue;
+			}
+
+			if (PartitionMethod(fkeyRelationid) == DISTRIBUTE_BY_NONE)
+			{
+				fkeyShardId = GetFirstShardId(fkeyRelationid);
+			}
+			else if (cacheEntry->partitionMethod == DISTRIBUTE_BY_HASH &&
+					 PartitionMethod(fkeyRelationid) == DISTRIBUTE_BY_HASH)
+			{
+				/* hash distributed tables should be colocated to have fkey */
+				Assert(TableColocationId(fkeyRelationid) == cacheEntry->colocationId);
+
+				fkeyShardId =
```

Comment:
there is a bug with using `fkeyRelationShard->relationId`. It is fixed on 2769, should be `fkeyRelationid `. I didn't want to do one more round of rebase just because of this, please manually change this line to test this PR, or maybe review 2769 first and look in detail here later.